From a6cc2ca5e322b39220c526bc83ea70d4c9230b2d Mon Sep 17 00:00:00 2001
From: Marcelo Vanzin <vanzin@cloudera.com>
Date: Mon, 28 Mar 2016 14:55:10 -0700
Subject: [PATCH 182/201] CLOUDERA-BUILD. CDH-38784. Fix ExternalSorter when
 shuffle encryption is enabled.

This fixes an issue in the original change for SPARK-5682 that was backported
to CDH 5.7.0.

(cherry picked from commit 7c720b13d482025eac2e4d7cecef9a40f4b45eec)
---
 .../spark/util/collection/ExternalSorter.scala     |   10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala b/core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala
index 44b1d90..6ab0e5f 100644
--- a/core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala
+++ b/core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala
@@ -26,6 +26,7 @@ import scala.collection.mutable
 import com.google.common.io.ByteStreams
 
 import org.apache.spark._
+import org.apache.spark.crypto.{CryptoConf, CryptoStreamUtils}
 import org.apache.spark.memory.TaskMemoryManager
 import org.apache.spark.serializer._
 import org.apache.spark.executor.ShuffleWriteMetrics
@@ -503,7 +504,14 @@ private[spark] class ExternalSorter[K, V, C](
           ", batchOffsets = " + batchOffsets.mkString("[", ", ", "]"))
 
         val bufferedStream = new BufferedInputStream(ByteStreams.limit(fileStream, end - start))
-        val compressedStream = blockManager.wrapForCompression(spill.blockId, bufferedStream)
+
+        val sparkConf = SparkEnv.get.conf
+        val maybeEncryptedStream = if (CryptoConf.isShuffleEncryptionEnabled(sparkConf)) {
+          CryptoStreamUtils.createCryptoInputStream(bufferedStream, sparkConf)
+        } else {
+          bufferedStream
+        }
+        val compressedStream = blockManager.wrapForCompression(spill.blockId, maybeEncryptedStream)
         serInstance.deserializeStream(compressedStream)
       } else {
         // No more batches left
-- 
1.7.9.5

